File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationExpect.kt
package com.github.dedis.popstellar.model.network.method.message.data.federation

import com.github.dedis.popstellar.model.network.method.message.MessageGeneral
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Data
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.google.gson.annotations.SerializedName

/** Federation Expect message */
class FederationExpect
/**
 * Constructor for a data Federation Expect
 *
 * @param laoId ID of the remote LAO
 * @param serverAddress public address of the remote organizer server
 * @param publicKey public key of the remote organizer
 * @param challenge challenge for the server
 */
(
    @SerializedName("lao_id") val laoId: String,
    @SerializedName("server_address") val serverAddress: String,
    @SerializedName("public_key") val publicKey: String,
    val challenge: MessageGeneral
) : Data {

  override val `object`: String
    get() = Objects.FEDERATION.`object`

  override val action: String
    get() = Action.EXPECT.action

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as FederationExpect
    return laoId == that.laoId &&
        serverAddress == that.serverAddress &&
        publicKey == that.publicKey &&
        challenge == that.challenge
  }

  override fun hashCode(): Int {
    return java.util.Objects.hash(laoId, serverAddress, publicKey, challenge)
  }

  override fun toString(): String {
    return "FederationExpect{lao_id='$laoId', server_address='$serverAddress'," +
        "public_key='$publicKey', challenge='$challenge'}"
  }
}

File: fe1-web/src/features/linked-organizations/network/messages/FederationExpect.ts
import { ActionType, Message, MessageData, ObjectType } from 'core/network/jsonrpc/messages';
import { validateDataObject } from 'core/network/validation';
import { Hash, ProtocolError, PublicKey } from 'core/objects';

/** Data sent to expect Federation */
export class FederationExpect implements MessageData {
  public readonly object: ObjectType = ObjectType.FEDERATION;

  public readonly action: ActionType = ActionType.FEDERATION_EXPECT;

  // The timestamp
  public readonly lao_id: Hash;

  public readonly server_address: String;

  public readonly public_key: PublicKey;

  public readonly challenge: Message;

  constructor(msg: Partial<FederationExpect>) {
    if (!msg.lao_id) {
      throw new ProtocolError("Undefined 'lao_id' parameter encountered during 'FederationExpect'");
    }
    if (!msg.server_address) {
      throw new ProtocolError(
        "Undefined 'server_address' parameter encountered during 'FederationExpect'",
      );
    }
    if (!msg.public_key) {
      throw new ProtocolError(
        "Undefined 'public_key' parameter encountered during 'FederationExpect'",
      );
    }
    if (!msg.challenge) {
      throw new ProtocolError(
        "Undefined 'challenge' parameter encountered during 'FederationExpect'",
      );
    }
    this.lao_id = msg.lao_id;
    this.server_address = msg.server_address;
    this.public_key = msg.public_key;
    this.challenge = msg.challenge;
  }

  /**
   * Creates an FederationExpect object from a given object
   * @param obj
   */
  public static fromJson(obj: any): FederationExpect {
    const { errors } = validateDataObject(ObjectType.FEDERATION, ActionType.FEDERATION_EXPECT, obj);
    if (errors !== null) {
      throw new ProtocolError(`Invalid federation expect\n\n${errors}`);
    }

    return new FederationExpect({
      lao_id: obj.lao_id,
      server_address: obj.server_address,
      public_key: obj.public_key,
      challenge: obj.challenge,
    });
  }
}
